core/common/scale.cpp

```diff
@@ -0,0 +1,506 @@
+          return expected::Value{ByteArray{minor_byte, major_byte}};
+        }
+
+        return encodeUint32(static_cast<uint32_t>(value));
+      }
+
+      EncodeIntegerResult encodeInt16(int16_t value) noexcept {
+        return encodeUint16(util::makeSignReinterpreter(value).asUnsigned());
+      }
+
+      EncodeIntegerResult encodeUint32(uint32_t value) noexcept {
+        if (value < EncodingCategoryLimits::kLessThanTwoDegreeThirty) {
+          uint32_t value_le = impl::encodeLittleEndian((value << 2) + 2);
+
+          impl::AsArray<uint32_t> arr{value_le};
+
+          return expected::Value{ByteArray(arr.begin(), arr.end())};
+        }
+
+        return encodeUint64(static_cast<uint64_t>(value));
+      }
+
+      EncodeIntegerResult encodeInt32(int32_t value) noexcept {
+        return encodeUint32(util::makeSignReinterpreter(value).asUnsigned());
+      }
+
+      EncodeIntegerResult encodeUint64(uint64_t value) noexcept {
+        return encodeBigInteger(BigInteger{value});
+      }
+
+      EncodeIntegerResult encodeInt64(int64_t value) noexcept {
+        if (value < 0)
+          return expected::Error{EncodeBigIntegerError::kBigNegativeValue};
+
+        return encodeUint64(static_cast<uint64_t>(value));
+      }
+
+      EncodeIntegerResult encodeBigInteger(const BigInteger &value) noexcept {
+        // cannot encode negative numbers
+        if (value < 0)
+          return expected::Error{EncodeBigIntegerError::kBigNegativeValue};
+
+        // calculate number of bytes required
+        auto countBytes = [](BigInteger v) -> size_t {
+          if (v == 0)
+            return 1;
+
+          size_t counter = 0;
+          while (v > 0) {
+            ++counter;
+            v /= 256;
+          }
+
+          return counter;
+        };
+
+        // number of bytes required to represent value
+        size_t bigIntLength = countBytes(value);
+
+        // number of bytes to scale-encode value
+        // 1 byte is reserved for header
+        size_t requiredLength = 1 + bigIntLength;
+
+        // not sure return error or just encode as is
+        // bigIntLength - 4 is encoded in value
+        if (bigIntLength < 4)
+          bigIntLength = 4;
+
+        if (bigIntLength > 67) {
+          return expected::Error{EncodeBigIntegerError::kValueIsTooBig};
+        }
+
+        ByteArray result;
+        result.reserve(requiredLength);
+        uint8_t header = (bigIntLength - 4) * 4 + 3;
+        result.push_back(header);
+        BigInteger v{value};
+        for (size_t i = 0; i < bigIntLength; ++i) {
+          result.push_back(static_cast<uint8_t>(v % 256));
+          v >>= 8;
+        }
+
+        return expected::Value{result};
+      }
+
+      Decoder::Decoder(const std::vector<uint8_t> &source)
+          : source_{source},
+            current_iterator_{std::begin(source_)},
+            bytes_left_{static_cast<int64_t>(source_.size())} {}
+
+      uint8_t Decoder::nextByte() {
```

Comment:
I think it's better check "hasMore" here and return optional

Reply:
Agree.